File: DevExpress-Demo/VB/MVVMDemo.Wpf/Modules/Services/DialogServiceViewModel.vb
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.POCO
Imports System.Threading.Tasks
Imports System.Windows

Namespace MVVMDemo.Services

    Public Class DialogServiceViewModel

        Public Sub ShowDetail(ByVal serviceName As String)
            Dim detailViewModel As DialogServiceDetailViewModel = DialogServiceDetailViewModel.Create()
            Dim registerCommand As UICommand = New UICommand() With {.Caption = "Register", .IsDefault = True, .Command = New DelegateCommand(Sub()
            End Sub, Function() Not String.IsNullOrEmpty(detailViewModel.CustomerName))}
            Dim cancelCommand As UICommand = New UICommand() With {.Caption = "Cancel", .IsCancel = True}
            Dim service As IDialogService = GetService(Of IDialogService)(serviceName)
            Dim result As UICommand = service.ShowDialog(dialogCommands:={registerCommand, cancelCommand}, title:="Registration Dialog", viewModel:=detailViewModel)
            If result Is registerCommand Then MessageBox.Show("Registered: " & detailViewModel.CustomerName)
        End Sub

        Public Sub ShowDetailAsync(ByVal serviceName As String)
            Dim detailViewModel As DialogServiceDetailViewModel = DialogServiceDetailViewModel.Create()
            Dim registerCommand As UICommand = New UICommand() With {.Caption = "Register", .IsDefault = True, .Command = New AsyncCommand(AddressOf Delay, Function() Not String.IsNullOrEmpty(detailViewModel.CustomerName)), .AsyncDisplayMode = AsyncDisplayMode.Wait}
            Dim cancelCommand As UICommand = New UICommand() With {.Caption = "Cancel", .IsCancel = True}
            Dim service As IDialogService = GetService(Of IDialogService)(serviceName)
            Dim result As UICommand = service.ShowDialog(dialogCommands:={registerCommand, cancelCommand}, title:="Registration Dialog", viewModel:=detailViewModel)
            If result Is registerCommand Then MessageBox.Show("Registered: " & detailViewModel.CustomerName)
        End Sub

        Private Async Function Delay() As Task
            Await Task.Delay(3000)
        End Function
    End Class
End Namespace

File: MyDX-Demo/MyDX-Demo/Module/MVVM-Demo/ServicesDemo/DialogServiceViewModel.cs
using DevExpress.Mvvm;
using DevExpress.Mvvm.POCO;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows;

namespace MyDX_Demo.Module.MVVM_Demo.ServicesDemo
{
    public class DialogServiceViewModel
    {
        public void ShowDetail(string serviceName)
        {
            DialogServiceDetailViewModel detailViewModel = DialogServiceDetailViewModel.Create();

            UICommand registerCommand = new UICommand()
            {
                Caption = "Register",
                IsDefault = true,
                Command = new DelegateCommand(() => { }, () => !string.IsNullOrEmpty(detailViewModel.CustomerName))
            };
            UICommand cancelCommand = new UICommand()
            {
                Caption = "Cancel",
                IsCancel = true,
            };
            IDialogService service = this.GetService<IDialogService>(serviceName);
            UICommand result = service.ShowDialog(
                dialogCommands: new[] { registerCommand, cancelCommand },
                title: "Registration Dialog",
                viewModel: detailViewModel
            );
            if (result == registerCommand)
                MessageBox.Show("Registered: " + detailViewModel.CustomerName);
        }
        public void ShowDetailAsync(string serviceName)
        {
            DialogServiceDetailViewModel detailViewModel = DialogServiceDetailViewModel.Create();

            UICommand registerCommand = new UICommand()
            {
                Caption = "Register",
                IsDefault = true,
                Command = new AsyncCommand(Delay, () => !string.IsNullOrEmpty(detailViewModel.CustomerName)),
                AsyncDisplayMode = AsyncDisplayMode.Wait
            };
            UICommand cancelCommand = new UICommand()
            {
                Caption = "Cancel",
                IsCancel = true,
            };

            IDialogService service = this.GetService<IDialogService>(serviceName);
            UICommand result = service.ShowDialog(
                dialogCommands: new[] { registerCommand, cancelCommand },
                title: "Registration Dialog",
                viewModel: detailViewModel
            );

            if (result == registerCommand)
                MessageBox.Show("Registered: " + detailViewModel.CustomerName);
        }

        async Task Delay()
        {
            await Task.Delay(3000);
        }
    }
}
